lake/journal/store.go

```diff
@@ -115,12 +124,70 @@ func (s *Store) load(ctx context.Context) error {
 			table[key] = e.Entry
 		case *Delete:
 			delete(table, e.EntryKey)
-		default:
-			return fmt.Errorf("unknown type in journal store: %T", e)
 		}
 	}
 }
 
+func (s *Store) getSnapshot(ctx context.Context) (ID, map[string]Entry, error) {
+	table := make(map[string]Entry)
+	r, err := s.journal.engine.Get(ctx, s.snapshotURI())
+	if err != nil {
+		return Nil, table, err
+	}
+	defer r.Close()
+	zr := zngio.NewReader(zed.NewContext(), r)
+	defer zr.Close()
+	v, err := zr.Read()
+	if v == nil || err != nil {
+		return Nil, table, err
+	}
+	if v.Type.ID() != zed.IDUint64 {
+		return Nil, table, errors.New("corrupted journal snapshot")
+	}
+	at := ID(zed.DecodeUint(v.Bytes))
+	for {
+		v, err := zr.Read()
+		if v == nil || err != nil {
+			return at, table, err
+		}
+		var e Entry
+		if err := s.unmarshaler.Unmarshal(v, &e); err != nil {
+			return at, nil, err
+		}
+		table[e.Key()] = e
+	}
+}
+
+func (s *Store) putSnapshot(ctx context.Context, at ID, table map[string]Entry) error {
+	// XXX This needs to be an atomic write for file systems.
+	w, err := s.journal.engine.Put(ctx, s.snapshotURI())
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+	zw := zngio.NewWriter(w)
```

Comment:
We should call zw.Close on error and let that call w.Close.
```suggestion
	zw := zngio.NewWriter(w)
	defer zw.Close()
```

Reply:
Tangential to this pr but, it's confusing that the zngio.Writer will close the writer passed to it but the reader will not.